text/0000-autoref-copy.md

```diff
@@ -0,0 +1,199 @@
+
+```rust
+// Comparisons:
+let v = vec![0, 1, 2, 3];
+v.iter().filter(|x| x > 1); // ERROR: expected &&_, found integral variable
+// These work:
+0 > 1;
+&0 > &1;
+// But these don't:
+&0 > 1;
+0 > &1;
+
+// Trait instantiations:
+let mut map: HashMap::new();
+map.insert(0, "Hello");
+map.insert(1, "world!");
+map[0]; // ERROR: expected &{integer}, found integral variable
+// or
+map.get(1); // ERROR: expected `&{integer}`, found integral variable
+
+// Numeric operators:
+// These work:
+&0 + &1;
+0 + &1;
+&0 + 1;
+// But these don't:
+&&0 + 1;
+&&0 + &1;
+&&0 + &&1;
+```
+
+These interactions confuse both new and experienced users without providing
+any significant value. It's clear what's intended by `map.get(1)` or
+`vec.iter().filter(|x| x > 1)`. When users encounter these errors in practice,
+the only reasonable thing they can do is add `&` and `*` as necessary to make
+their code compile.
+
+This RFC seeks to address one particular variant of this problem: passing
+owned `Copy` data where a reference was expected.
+
+Example:
+
+```rust
+use std::collections::HashMap;
+fn main() {
+    let mut map = HashMap::new();
+    map.insert(1, "hello!");
+
+    println!("{:?}", map.get(1)); // ERROR: expected `&{integer}`
+
+    // Instead, users have to write this:
+    println!("{:?}", map.get(&1));
+}
+```
+
+This is an unnecessary frustration. This RFC proposes to prevent this issue
+by auto-referencing `Copy` types.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When calling a function which expects a reference to a `Copy` type, you can
+pass an owned `Copy` type and a reference will be created automatically:
+
+```rust
+fn print_num(x: &usize) { println!("{:?}", x); }
+
+fn main() {
+  let x = 5;
+  print_num(x);
+}
+```
+
+This is particularly convenient when working with some generic types whose
+methods take references as arguments:
+
+```rust
+use std::collections::HashMap;
+fn main() {
+    let mut map = HashMap::new();
+    map.insert(1, "hello!");
+
+    // We can write this:
+    println!("{:?}", map.get(1));
+
+    // Instead of having to write this:
+    println!("{:?}", map.get(&1));
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`Copy` types autoreference: at
+[coercion sites](https://github.com/rust-lang/rfcs/blob/master/text/0401-coercions.md#coercions),
+when a reference is expected, but an owned
+`Copy` type is found, an `&` will be automatically inserted.
+
+Some types have interior mutability through `UnsafeCell`, so passing them
+via a hidden reference could allow surprising mutations to occur.
+To prevent this, the autoreferencing coercion is limited to types which do not
+directly contain an internal `UnsafeCell`.
+Types which reference an `UnsafeCell` through an indirection can still be
+coerced (e.g. `&MyCopyCell<u8>` to `&&MyCopyCell<u8>`).
+
+This coercion will not occur if the lifetime of the resulting reference would
```

Comment:
That... sounds like a good way to cause "inference ripping itself apart" issues, at the very least. Type-checking is necessarily independent of region inference, and even more necessarily independent of it with NLL.

Reply:
Does this need a dependency upon region inference? I'd expect it'd only need to examine the signature of the function being called to check if the argument lifetime either comes from outside the function or escapes via the return type.

Reply:
@cramertj On function calls perhaps, but function calls are not the only coercion site.

@arielb1  I'm a noob on compiler internals, but perhaps the type checker could coerce ignoring the lifetime restriction and pass the restriction down to borrowck which would emit a "plz type `&` here" error if the lifetime is too long.

Reply:
@cramertj 

Say hi to generic types, associated types, etc.